lib/capybara-lockstep/capybara_ext.rb

```diff
@@ -3,36 +3,73 @@
 module Capybara
   module Lockstep
     module SynchronizeMacros
+      # Helper module to avoid unnecessary double synchronization in case we patched a class and a subclass, i.e.:
+      # Capybara::Selenium::ChromeNode#click should run Lockstep.auto_synchronize, but it also might super to
+      # Capybara::Selenium::Node#click which should not also run Lockstep.auto_synchronize.
+      # However, we have to patch both classes, in case Capybara::Selenium::Node is used directly.
+      module Nestable
+        def self.skip_nested_calls(identifier, &block)
+          thread_local_key = self.thread_local_key(identifier)
+          if Thread.current[thread_local_key]
+            yield
+          else
+            begin
+              Thread.current[thread_local_key] = true
+              yield
+            ensure
+              Thread.current[thread_local_key] = nil
+            end
+          end
+        end
+
+        def self.nested_call?(identifier)
+          Thread.current[thread_local_key(identifier)]
+        end
+
+        private
+
+        def self.thread_local_key(identifier)
+          :"Capybara::Lockstep_skip_#{identifier}"
+        end
+      end
+
+      def self.extended(by)
+        by.instance_eval do
+          prepend(@before_synchronize_module = Module.new)
+          prepend(@synchronize_after_module = Module.new)
+          prepend(@unsynchronize_after_module = Module.new)
+        end
+      end
 
       def synchronize_before(meth, lazy:)
-        mod = Module.new do
+        @before_synchronize_module.module_eval do
           define_method meth do |*args, &block|
-            Lockstep.auto_synchronize(lazy: lazy, log: "Synchronizing before ##{meth}")
-            super(*args, &block)
+            Lockstep.auto_synchronize(lazy: lazy, log: "Synchronizing before ##{meth}") unless Nestable.nested_call?(:synchronize_before)
+            Nestable.skip_nested_calls(:synchronize_before) do
+              super(*args, &block)
+            end
           end
 
           ruby2_keywords meth
         end
-
-        prepend(mod)
       end
 
       def synchronize_after(meth)
-        mod = Module.new do
+        @synchronize_after_module.module_eval do
           define_method meth do |*args, &block|
-            super(*args, &block)
+            Nestable.skip_nested_calls(:synchronize_after) do
+              super(*args, &block)
+            end
           ensure
-            Lockstep.auto_synchronize
+            Lockstep.auto_synchronize(log: "Synchronizing after ##{meth}") unless Nestable.nested_call?(:synchronize_after)
```

Comment:
Finde der Kontrollfluss hat sehr viele bewegliche Teile: `Nestable` + `Thread.current` + Methode muss 2x mit `Nested` sprechen.

Wäre hier eine weniger smarte Lösung einfacher nachzuvollziehen? Sowas wie:

```ruby
def synchronize_after(meth)
  prepend Module.new do
    define_method meth do |*args, &block|
      @synchronize_after_calls ||= 0
      @synchronize_after_calls += 1
      super(*args, &block)
    ensure
      Lockstep.auto_synchronize(log: "Synchronizing after ##{meth}") if @synchronize_after_calls == 1
      @synchronize_after_calls -= 1
    end
  end
end  
```

Alternativ fände ich es auch gut wenn sich die Lockstep-Klasse um das Halten des "Locks" (`@synchronize_after_call`) Thread-lokal kümmert.

```ruby
def synchronize_after(meth)
  prepend Module.new do
    define_method meth do |*args, &block|
      Lockstep.auto_synchronize_after(log: "Synchronizing after ##{meth}") do
        super(*args, &block)
      end  
    end
  end
end  
```

Reply:
Zweite Lösung find ich nicht so toll, denke dass die Logik nicht in die Lockstep-Klasse reinsollte.

Hab jetzt den ersten Weg umgesetzt, das ist sicher einen Tick einfacher. Wollte vermeiden Instanz-Variablen in fremden Objekten zu hinterlassen, aber glaube es ist so wirklich besser.